### PR TITLE
Do not allow await in catch/finally in language version 5

### LINKED ISSUE #2725: Await in catch and finally accepted in langversion 5
The Roslyn C# compiler accepts await in catch and finally blocks even in language version 5. It should be an error unless in language version 6 or higher.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #2725 

Reviewers: @VSadov @jaredpar @AlekseyTs @agocke

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs [hunk 1]
```diff
@@ -118,32 +118,42 @@ private bool ReportBadAwaitWithoutAsync(CSharpSyntaxNode node, DiagnosticBag dia
...
             }
             else if (this.Flags.Includes(BinderFlags.InCatchFilter))
             {
                 Error(diagnostics, ErrorCode.ERR_BadAwaitInCatchFilter, node);
-                hasErrors = true;
+                return true;
+            }
+            else if (this.Flags.Includes(BinderFlags.InFinallyBlock) &&
+                (node.SyntaxTree as CSharpSyntaxTree)?.Options?.IsFeatureEnabled(MessageID.IDS_AwaitInCatchAndFinally) == false)
+            {
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: Nice!

**Comment by VSadov** [comment]
LGTM

**Comment by jaredpar** [comment]
:+1:

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs [hunk 2]
```diff
@@ -118,32 +118,42 @@ private bool ReportBadAwaitWithoutAsync(CSharpSyntaxNode node, DiagnosticBag dia
...
+                return true;
             }
             else if (this.Flags.Includes(BinderFlags.InCatchFilter))
             {
                 Error(diagnostics, ErrorCode.ERR_BadAwaitInCatchFilter, node);
-                hasErrors = true;
+                return true;
+            }
+            else if (this.Flags.Includes(BinderFlags.InFinallyBlock) &&
+                (node.SyntaxTree as CSharpSyntaxTree)?.Options?.IsFeatureEnabled(MessageID.IDS_AwaitInCatchAndFinally) == false)
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: Can node.SyntaxTree not be a CSharpSyntaxTree?

Comment (tmat) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: Also Options should always be non-null, no?

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: We don't have complete control of how trees can be created now or in the future. If the usual APIs are used (e.g. Parse) the null checks aren't necessary. This is mainly defensive.

Comment (tmat) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: I am not a fan of such defensive measures. Options should never be null. There are many places in the source base where we don't guard against null Options. A better defensive measure would be to change SyntaxTree.Options as follows:

```C#
public ParseOptions Options
        {
            get
            {
                var options = this.OptionsCore;
                if (options == null) throw new InvalidOperationException();
                return options;
            }
        }
```

Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs]: @tmat I think it would be fine to strengthen the preconditions for trees passed to the semantic model, but that is outside the scope of this change.
